### PR TITLE
Refactor forum forms

### PR DESCRIPTION
**Pull request by gasche**
Après #5082 qui retirait des redondances de `zds/forum/feeds`, cette PR factorise du code de vérification de champs dans `zds/forum/forms`, en mettant du code utilitaire dans `zds/utils/forms` -- peut-être utilisable en dehors de `zds/forum` à terme ?

Cette PR dépend de #5083 (donc les commits de #5083 sont aussi inclus ici), dont j'ai eu besoin pour écrire un nouveau fichier de test `zds/forum/tests/tests_forms`. Ces tests ne dépendent pas d'avoir une instance du serveur qui tourne, donc j'ai pu tester, et les nouveaux tests passent.

### CODE REVIEW

File: zds/forum/forms.py [hunk 1]
```diff
@@ -7,10 +7,10 @@
 from crispy_forms.layout import Layout, Field, Hidden, HTML
 from crispy_forms.bootstrap import StrictButton
 from zds.forum.models import Forum, Topic
-from zds.utils.forms import CommonLayoutEditor, TagValidator
+from zds.utils.forms import CommonLayoutEditor, TagValidator, FieldValidatorMixin
 
 
-class TopicForm(forms.Form):
+class TopicForm(forms.Form, FieldValidatorMixin):
```
Comment (artragis) [inline zds/forum/forms.py:13]: je me demande à quel point il ne faudrait pas utiliser des ModelForm

File: zds/forum/forms.py [hunk 2]
```diff
@@ -77,34 +77,24 @@ def __init__(self, *args, **kwargs):
     def clean(self):
         cleaned_data = super(TopicForm, self).clean()
 
-        title = cleaned_data.get('title')
-        text = cleaned_data.get('text')
-        tags = cleaned_data.get('tags')
+        self.get_non_empty_or_error(cleaned_data, 'title',
+                                    lambda: _('Le champ titre ne peut être vide'))
+        text = self.get_non_empty_or_error(cleaned_data, 'text',
```
Comment (artragis) [inline zds/forum/forms.py]: `get_not_empty_field_or_error`

**Review by artragis — CHANGES_REQUESTED** [review]

File: zds/utils/forms.py [hunk 3]
```diff
@@ -143,3 +143,19 @@ def validate_no_empty_slug(self, tag):
     @property
     def errors(self):
         return self.__errors
+
+
+class FieldValidatorMixin(object):
```
Comment (artragis) [inline zds/utils/forms.py]: no need of `object` inheritance, we do not want to be compatible with py2

File: zds/forum/forms.py (hunk 1 above)
Comment (gasche) [reply zds/forum/forms.py:13]: Je ne connais pas les ModelForm, donc je vais passer mon tour. Je me contente de refactorer du code existant pour réduire un peu les redondances.

File: zds/forum/forms.py (hunk 2 above)
Comment (gasche) [reply zds/forum/forms.py]: Merci, j'ai renommé. J'ai aussi changé le nom de l'autre méthode pour rajouter `text` dans le nom, puisque le champ `errors['text']` est hardcodé dans son implémentation.

File: zds/utils/forms.py (hunk 3 above)
Comment (gasche) [reply zds/utils/forms.py]: Done, thanks.

**Comment by gasche** [comment]
(Maintenant que #5082 est mergé, y a-t-il des plans pour celle-ci ?)

**Comment by artragis** [comment]
mon plan c'est de faire une relecture complète. Mais je te conseille d'attendre un peu avant de rebase car vu que la PR de @Eskimon sur les stats est prête elle va arriver avant la tienne.

**Comment by artragis** [comment]
tu peux rebase celle-là stp? (c'est celle-ci que je mergerai en premier)

**Comment by gasche** [comment]
C'est rebasé.

File: zds/forum/forms.py [hunk 4]
```diff
@@ -77,34 +77,24 @@ def __init__(self, *args, **kwargs):
...
-        if title is not None:
-            if not title.strip():
-                self._errors['title'] = self.error_class(
-                    [_('Le champ titre ne peut être vide')])
-                if 'title' in cleaned_data:
-                    del cleaned_data['title']
-        if text is not None and not text.strip():
-            self._errors['text'] = self.error_class(
-                [_('Le champ text ne peut être vide')])
-            if 'text' in cleaned_data:
```
Comment (artragis) [inline zds/forum/forms.py:93]: pourquoi avoir enlevé cette partie?

File: zds/forum/tests/tests_forms.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        form = TopicForm(data=data)
+
+        self.assertFalse(form.is_valid())
+
+    def test_invalid_topic_form_text_too_long(self):
+        """ Test when text runs over the length limit """
+        data = {
+            'title': 'Test Topic Title',
+            'subtitle': 'Test Topic Subtitle',
+            'tags': 'test,topic,tags',
```
Comment (artragis) [inline zds/forum/tests/tests_forms.py:87]: can you add test with empty tags (must be valid)

**Review by artragis — CHANGES_REQUESTED** [review]

File: zds/forum/forms.py (hunk 4 above)
Comment (gasche) [reply zds/forum/forms.py:93]: De quelle partie parles-tu ? Il me semble que globalement ce code retiré (des lignes 84 à 93) est rendu redondant par les nouvelles lignes 80-81 (pour les anciennes 84-89), 82-83 (pour 90-94) et 85-88 (pour 96-99).

Comment (artragis) [reply zds/forum/forms.py:93]: en fait c'est le fait qu'il n'y ait pas d'équivalent au `del cleaned_data['text']` qui me fausait poser la question.

En soi la logique d'affichage des erreurs est respectée mais dans l'ancien code on supprimait le champ `text`, peut être est-ce inutile et donc tu auras eu raison d'enlever cette partie, mais peut-être y a-t-il une raison.

Comment (gasche) [reply zds/forum/forms.py:93]: La fonction utilitaire `get_non_empty_field_or_error` que j'ai définie fait bien ce `del cleaned_data[key]` (ligne 156 de forms.py). Du coup, ma factorisation introduit d'ailleurs des suppressions à des endroits où il n'y en avait pas avant, dans le cas du champ `'text'` de la classe `PostForm` ci-dessous. Le code entre les deux classes était incohérent, j'imagine qu'il faut retirer soit partout soit nulle part, et j'ai fait ce choix-là. L'idée est que `cleaned_data` ne contient que les valeurs valides à la fin de l'étape de validation.

File: zds/forum/tests/tests_forms.py (hunk 5 above)
Comment (gasche) [reply zds/forum/tests/tests_forms.py:87]: done

File: zds/forum/forms.py (hunk 4 above)
Comment (artragis) [reply zds/forum/forms.py:93]: Impeccable !

File: zds/forum/tests/tests_forms.py [hunk 6]
```diff
@@ -8,7 +8,6 @@
 
 text_too_long = (settings.ZDS_APP['forum']['max_post_length'] + 1) * 'a'
 
-
```
Comment (artragis) [inline zds/forum/tests/tests_forms.py]: fallait pas retirer cette ligne-là.

**Review by artragis — CHANGES_REQUESTED** [review]

File: zds/forum/tests/tests_forms.py (hunk 6 above)
Comment (gasche) [reply zds/forum/tests/tests_forms.py]: (grmbl, grbml. Autant j'aime bien l'indentation obligatoire, autant je trouve que les contraintes sur le nombre de lignes vides entre deux machins c'est vraiment chiant.) Je corrige.

Comment (artragis) [reply zds/forum/tests/tests_forms.py]: en fait ça c'est pep8 qui le demande (le linter) c'est pour avoir un code uniformisé.

Comment (gasche) [reply zds/forum/tests/tests_forms.py]: J'entends bien, mais il me semble que cette uniformité n'a pas grand intérêt. Enfin j'imagine que pour les gens qui écrivent en Python régulièrement c'est naturel et ça ne les dérange plus, mais pour des contributeurs occasionnels c'est chiant. (Ce n'est pas une critique de zds, mais de la PEP.)
